packages/backend/src/services/test-run.ts

```diff
@@ -1,11 +1,129 @@
 import { type IAction } from '@plumber/types'
 
 import { getOnPipePublishOrBeforeTestRunHook } from '@/helpers/actions'
+import { isIfThenStep } from '@/helpers/toolbelt'
 import Step from '@/models/step'
 import { processAction } from '@/services/action'
 import { processFlow } from '@/services/flow'
 import { processTrigger } from '@/services/trigger'
 
+/**
+ *
+ * == EDGE CASE ==
+ *
+ * If we are testing a step within an if-then branch, we only want to execute
+ * steps within that branch, as well as all steps located in ancestor branches.
+ *
+ * ---
+ * Example:
+ *
+ * [T = trigger S = normal step, B = branch]
+ *
+ *   T
+ *   |
+ *   S1
+ *   |
+ *   B1
+ *     \
+ *      B1.1
+ *        \
+ *         S2
+ *         |
+ *         S3
+ *        /
+ *      B1.2
+ *        \
+ *         S4
+ *       /
+ *      /
+ *   B2
+ *     \
+ *      S5
+ *       |
+ *      B2.1
+ *        \
+ *         S6
+ *        /
+ *      B2.2
+ *        \
+ *         S7
+ *
+ * If we are testing step S7, then we only want to execute:
+ * - Concrete steps: S1, S5 and S7.
+ * - Branch and sibling branch steps: B1, B2, B2.1 and B2.2
+ *   - This is so that we can tell if S7 would have been skipped.
+ *   - Running all sibling branches is needed because we need to keep nextStepId
+ *     consistent in our main execution loop.
+ * ---
```

Comment:
can i just check if we run B2 or S5, is it only supposed to run S1 only? Cus I think now it will run S4 also right

Reply:
Eh paiseh ignore this. the old logic is broken

Reply:
anyway its draft... i dont think i want to merge this in at all unless there's a strong callout. Too tricky

Reply:
but yes you pointed out a bug lol, its broken if we run branch steps.